### DISCUSSION TITLE
New samplers "ipndm" and "ipndm_v"; what's the difference?

### DISCUSSION THREAD

**Discussion by not-ski** [comfyanonymous/ComfyUI#3799] [Q&A]
I saw that "ipndm" and "ipndm_v" samplers were added (https://github.com/comfyanonymous/ComfyUI/commit/d5efde89b76c40c72668daec052c07c71a737908); what exactly is the difference between these two?

From scouring the linked source repo, I gather that ipndm is a fourth order sampler and that ipndm_v is the "variable-step version". But what does variable-step mean here?

Thanks <3

**Comment by zhyzhouu** [comment]
iPNDM is actually the classical Adams–Bashforth methods, which is derived for a fixed step size (uniformly sampled sigmas). But for diffusion sampling, the stride of each step is usually different, so we also provide a variable-step version of iPNDM denoted as iPNDM_v. However, we find iPNDM empirically performs better especially for small-resolution datasets. For detailed derivation, you can refer to [Interval methods of Adams-Bashforth type with variable step sizes](https://link.springer.com/article/10.1007/s11075-019-00774-y).

**Reply by not-ski** [reply]
Thanks, that's helpful!

**Reply by simonlui** [reply]
@zhyzhouu I looked at the repo and paper you linked at https://github.com/zju-pi/diff-sampler/tree/main/gits-main but just to confirm, it seems like the table in the repo says you combine iPNDM with GITS for best results, right? GITS is a custom scheduler node right now, so we need iPNDM and iPNDM_v as seperate nodes as well. ~~I think it is only implemented as part of the standard Ksampler right now. I can write them as separate nodes hopefully in a quick PR.~~

Edit: Never mind, ComfyUI doesn't need that, you can just select the iPNDM samplers with the KSamplerSelect node, I forgot that was a thing.

**Reply by not-ski** [reply]
YMMV, but anecdotally/subjectively I only got _particularly_ good results out of GITS at <= 20 steps. At > 20 steps I generally preferred the results from AYS scheduler instead. This was using a finetune of Pony SDXL.

Also, with SDXL I was only able to consistently get usable results at <= 20 steps w/ GITS using DPM++ 2M or UniPC samplers.

IPNDM seemed to require higher step counts to achieve consistently usable/convergent results.

Again, this is all 100% anecdotal and subjective and not scientific by any means, but these were my observations at least.

**Reply by simonlui** [reply]
The [chart](https://raw.githubusercontent.com/zju-pi/diff-sampler/main/gits-main/assets/fid.png) posted in the repo I linked from the paper pegs the combination of GITS and IPNDM converging fast to good quality at 5 NFE which I think is equal to 10 steps? But yeah, I don't see anything for 20+ steps/10+ NFE but I assume the research didn't experiment with what happens to the image if you keep going beyond that limit.

**Reply by not-ski** [reply]
This could be a Pony thing, but I was only really able to get good results at 10 steps w/ GITS using DPM++ 2M. IPNDM definitely seemed to require quite a few more steps to converge (it wouldn't even do so consistently at 20 steps).

**Reply by simonlui** [reply]
These are academic benchmarks after all so yeah, I am not surprised if the way a model was trained or fine tuned also affects it. The coefficient might make a difference but I have no idea how that can be tuned per model unfortunately.

**Reply by zhyzhouu** [reply]
> The [chart](https://raw.githubusercontent.com/zju-pi/diff-sampler/main/gits-main/assets/fid.png) posted in the repo I linked from the paper pegs the combination of GITS and IPNDM converging fast to good quality at 5 NFE which I think is equal to 10 steps? But yeah, I don't see anything for 20+ steps/10+ NFE but I assume the research didn't experiment with what happens to the image if you keep going beyond that limit.

Indeed, GITS is designed for fast sampling of diffusion models, and we did not test large steps (also due to limited resources). As for NFE, it refers to the number of network evaluations to generate an image. Usually 1 NFE = 1 step. But for Stable Diffusion, since two network evaluations are required in a single step, 1NFE=2*steps in this case.

**Reply by zhyzhouu** [reply]
> This could be a Pony thing, but I was only really able to get good results at 10 steps w/ GITS using DPM++ 2M. IPNDM definitely seemed to require quite a few more steps to converge (it wouldn't even do so consistently at 20 steps).

Actually, higher-order solvers are not perferred by Stable Diffusion. Take DPM++ as example, though DPM++(3M) (third-order) exists, it performs worse than DPM++(2M) (second-order). The reason still needs to be studied. Currently, the iPNDM sampler is by default a fourth-order algorithm (max_order=4). If you change `max_order` from 4 to 2, iPNDM will have similar performance with DPM++(2M) (empirically verified in my early experiments).

**Reply by not-ski** [reply]
Very interesting! Yeah I have generally noticed that in Stable Diffusion, DPM++ 3M requires *many* more steps than 2M to achieve usable/convergent results.

Anecdotally, at very high step counts (50+) I have preferred the results from higher-order solvers. But certainly, when it comes to speed they seem to be empirically inferior.

**Reply by zhyzhouu** [reply]
> These are academic benchmarks after all so yeah, I am not surprised if the way a model was trained or fine tuned also affects it. The coefficient might make a difference but I have no idea how that can be tuned per model unfortunately.

Unfortunately, I am currently not sure how different fine-tuning will affect the model too. 🤯

**Reply by simonlui** [reply]
>Anecdotally, at very high step counts (50+) I have preferred the results from higher-order solvers. But certainly, when it comes to speed they seem to be empirically inferior.

It doesn't save that much time per step but total steps needed, agreed. My anecdotal experience having played around with iPNDM and SD3 tonight and tweaking the max_order, the coherence of the prompt to not generate extra limbs and etc. with human subject seems to go down if you lower the max order so I supposed there is some correlation there as well.

**Reply by not-ski** [reply]
I'd be interested to see what results of your empirical testing on SDXL. IME, ipndm_v has noticeably outperformed ipndm (using GITS or AYS and targeting 20 steps).

Both are still generally giving lesser quality results than lower-order solvers at that step count, but ipndm_v gets much closer to a coherent/convergent result.

This is odd, considering it directly contradicts the data gathered in your previous testing. I wonder if this is another quirk of Stable Diffusion.

**Reply by kalle07** [reply]
today in webforge-ui 4 new samplers
ipndm and ipndm_v
HeunPP2
DEIS
no use for FLUX  ... why   : D
